File: client/src/protocol.ts
export type PairingPayload = {
  v: number;
  endpoint: string;
  fingerprint: string;
  skipFingerprintVerification?: boolean;
  pairingId: string;
  token: string;
  expiresAt: string;
};

export type WaitState = {
  kind: string;
  label: string;
  confidence: number;
  matched: string;
};

export type SessionSummary = {
  id: string;
  name: string;
  command: string;
  cwd: string;
  state: 'running' | 'exited' | 'waiting' | 'idle';
  createdAt: string;
  updatedAt: string;
  preview: string[];
  waitState?: WaitState;
};

export type AgentModelInfo = {
  id: string;
  name: string;
  provider: string;
};

export type AgentCapability = { name: string; enabled: boolean };

export type AgentSession = {
  id: string;
  adapter: string;
  terminalSessionId: string;
  cwd: string;
  state: 'working' | 'idle' | 'needsYou' | 'exited';
  capabilities: AgentCapability[];
  model?: AgentModelInfo;
  thinking?: string;
  availableModels?: AgentModelInfo[];
  availableThinking?: string[];
  createdAt: string;
  updatedAt: string;
};
export type AgentEvent = {
  type: string;
  cursor?: number;
  eventId?: string;
  agentId: string;
  messageId?: string;
  toolCallId?: string;
  text?: string;
  toolName?: string;
  toolInput?: unknown;
  toolOutput?: unknown;
  state?: string;
  capabilities?: AgentCapability[];
  model?: AgentModelInfo;
  thinking?: string;
  availableModels?: AgentModelInfo[];
  availableThinking?: string[];
  isError?: boolean;
};

export type AgentHistoryResponse = {
  cursor: number;
  events: AgentEvent[];
};
export type RuntimeLifecycleEvent = {
  cursor?: number;
  surfaceId: string;
  type: string;
  payload: unknown;
};

export type TmuxPane = {
  terminalSessionId: string;
  serverId: string;
  sessionId: string;
  windowId: string;
  paneId: string;
  sessionName: string;
  windowName: string;
  windowIndex: number;
  paneIndex: number;
  cwd: string;
  active: boolean;
};

export type RuntimeSnapshot = {
  cursor: number;
  terminals: Array<{ id: string; name: string; cwd: string; seq: number; exited: boolean }>;
  agents: AgentSession[];
  topology: TmuxPane[];
  desktops: unknown[];
};

export type ChannelOpenEnvelope = {
  type: 'channel.open'; requestId: string; channelId: string; kind: string; targetId: string; after?: number;
};
export type ChannelCloseEnvelope = { type: 'channel.close'; channelId: string };
export type CommandEnvelope = { type: 'command'; requestId: string; targetId: string; command: string; args?: unknown };
export type ChannelOpenedEnvelope = { type: 'channel.opened'; requestId: string; channelId: string; cursor: number };
export type RuntimeEventEnvelope = { type: 'event'; channelId: string; cursor: number; event: unknown };
export type CommandResultEnvelope = { type: 'command.result'; requestId: string; ok: boolean; result?: unknown; error?: string };
export type ChannelClosedEnvelope = { type: 'channel.closed'; channelId: string; reason: string };
export type HostIdentity = {
  hostId: string;
  connectionId: string;
  sessionId: string;
};

export type Capability = {
  name: string;
  enabled: boolean;
};

export type CapabilitiesResponse = {
  capabilities: Capability[];
};

export type DaemonCapabilities = {
  identity: HostIdentity;
  capabilities: Capability[];
};


export type CreateSessionRequest = {
  name: string;
  command: string;
  args: string[];
  cwd: string;
  cols: number;
  rows: number;
  backend?: string;
};

export type FileEntry = {
  path: string;
  name: string;
  isDir: boolean;
  size: number;
  mode: string;
  gitCode?: string;
};

export type ReadFileResponse = { path: string; sha256: string; text: string };
export type ListFilesResponse = {
  entries: FileEntry[];
};

export type RenameFileRequest = { path: string; newPath: string };
export type CopyFileRequest = { path: string; newPath: string };

export type GitEntry = {
  code: string;
  path: string;
};

export type GitStatus = {
  available: boolean;
  entries: GitEntry[];
};

export type GitStatusResponse = {
  available: boolean;
  entries: GitEntry[];
};

export type ListShellsResponse = { shells: string[] };
export type NotifyRegisterRequest = {
  provider: string;
  token: string;
};

export type ErrorEnvelope = { type: 'error'; code: string; message: string };

export type DesktopSessionResponse = {
  ticket: string;
  wsUrl: string;
  expiresAt: string;
};

export type PTYInputEnvelope = {
  type: 'pty.input';
  sessionId: string;
  data: string;
};

export type PTYOutputEnvelope = {
  type: 'pty.output';
  sessionId: string;
  data: string;
  seq: number;
};

export type PTYResizeEnvelope = {
  type: 'pty.resize';
  sessionId: string;
  cols: number;
  rows: number;
};

export type SessionStateEnvelope = {
  type: 'session.state';
  sessionId: string;
  state: string;
  waitState?: WaitState;
};

File: backend/internal/protocol/protocol.go
package protocol

import "time"

type HealthResponse struct {
	OK      bool   `json:"ok"`
	Version string `json:"version"`
}

type SessionSummary struct {
	ID        string     `json:"id"`
	Name      string     `json:"name"`
	Command   string     `json:"command"`
	CWD       string     `json:"cwd"`
	State     string     `json:"state"`
	CreatedAt time.Time  `json:"createdAt"`
	UpdatedAt time.Time  `json:"updatedAt"`
	Preview   []string   `json:"preview"`
	WaitState *WaitState `json:"waitState"`
}

type WaitState struct {
	Kind       string  `json:"kind"`
	Label      string  `json:"label"`
	Confidence float64 `json:"confidence"`
	Matched    string  `json:"matched"`
}

type CreateSessionRequest struct {
	Name    string   `json:"name"`
	Command string   `json:"command"`
	Args    []string `json:"args"`
	CWD     string   `json:"cwd"`
	Cols    int      `json:"cols"`
	Rows    int      `json:"rows"`
}

type ResizeSessionRequest struct {
	Cols int `json:"cols"`
	Rows int `json:"rows"`
}

type InputRequest struct {
	Data string `json:"data"`
}

type WriteFileRequest struct {
	Path           string `json:"path"`
	Content        string `json:"content"`
	ExpectedSHA256 string `json:"expectedSha256"`
}

type RenameFileRequest struct {
	Path    string `json:"path"`
	NewPath string `json:"newPath"`
}

type AuthHello struct {
	Type        string `json:"type"`
	PairingID   string `json:"pairingId"`
	ClientNonce string `json:"clientNonce"`
	ClientName  string `json:"clientName"`
}

type AuthChallenge struct {
	Type        string `json:"type"`
	ServerNonce string `json:"serverNonce"`
	ChallengeID string `json:"challengeId"`
	Salt        string `json:"salt"`
}

type AuthProof struct {
	Type        string `json:"type"`
	PairingID   string `json:"pairingId"`
	ChallengeID string `json:"challengeId"`
	Proof       string `json:"proof"`
}

type AuthOK struct {
	Type         string `json:"type"`
	SessionToken string `json:"sessionToken"`
}

type AuthToken struct {
	Type  string `json:"type"`
	Token string `json:"token"`
}

type PTYInputEnvelope struct {
	Type      string `json:"type"`
	SessionID string `json:"sessionId"`
	Data      string `json:"data"`
}

type PTYOutputEnvelope struct {
	Type      string `json:"type"`
	SessionID string `json:"sessionId"`
	Data      string `json:"data"`
	Seq       int64  `json:"seq"`
}

type PTYResizeEnvelope struct {
	Type      string `json:"type"`
	SessionID string `json:"sessionId"`
	Cols      int    `json:"cols"`
	Rows      int    `json:"rows"`
}

type SessionStateEnvelope struct {
	Type      string     `json:"type"`
	SessionID string     `json:"sessionId"`
	State     string     `json:"state"`
	WaitState *WaitState `json:"waitState"`
}

type ErrorEnvelope struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

type FileEntry struct {
	Path    string `json:"path"`
	Name    string `json:"name"`
	IsDir   bool   `json:"isDir"`
	Size    int64  `json:"size"`
	Mode    string `json:"mode"`
	GitCode string `json:"gitCode,omitempty"`
}

type ListFilesResponse struct {
	Entries []FileEntry `json:"entries"`
}

type ReadFileResponse struct {
	Path   string `json:"path"`
	SHA256 string `json:"sha256"`
	Text   string `json:"text"`
}

type GitStatusResponse struct {
	Available bool       `json:"available"`
	Entries   []GitEntry `json:"entries"`
}

type GitEntry struct {
	Code string `json:"code"`
	Path string `json:"path"`
}

type NotifyRegisterRequest struct {
	Provider string `json:"provider"`
	Token string `json:"token"`
}

type ListShellsResponse struct {
	Shells []string `json:"shells"`
}
